GitUI/Avatars/AvatarService.cs

```diff
@@ -1,12 +1,132 @@
+            {
+                0 => null,
+                1 => providers[0],
+                _ => new ChainedAvatarProvider(providers),
+            };
+
+            // Local methods to build requested main and fallback providers:
+
+            IAvatarProvider BuildMainProvider()
+            {
+                return provider switch
+                {
+                    AvatarProvider.Default => BuildDefaultMainProvider(),
+                    AvatarProvider.Custom => CustomAvatarProvider.ParseTemplateString(AppSettings.CustomAvatarTemplate, lazyDownloader.Value),
+                    _ => null,
+                };
+            }
+
+            IAvatarProvider BuildDefaultMainProvider()
+            {
+                // 1. query github (with non-reply and regular email addresses)
+                //    github might internally fall back to gravatar, so only need a single request in most cases.
+
+                // 2. resolve via gravatar (for users that don't have a github account)
+                //    this request also directly provides the fallback if it's gravatar compatible.
+
+                return new ChainedAvatarProvider(
+                    new GithubAvatarProvider(lazyDownloader.Value),
+                    new GravatarProvider(lazyDownloader.Value, fallbackType));
+            }
+
+            IAvatarProvider BuildFallbackProvider()
+            {
+                return fallbackType switch
+                {
+                    // Local fallback providers
+                    AvatarFallbackType.MysteryMan => UserImageAvatarProvider,
+                    AvatarFallbackType.AuthorInitials => InitialsAvatarProvider,
+
+                    // Use gravatar as fallback provider
+                    var type when GravatarProvider.IsFallbackSupportedByGravatar(fallbackType.Value) => new GravatarProvider(lazyDownloader.Value, type, true),
+
+                    // Use author initials for unknown types
+                    _ => InitialsAvatarProvider,
+                };
+            }
+        }
+
+        /// <summary>
+        /// Wraps a persistance/cache layer arround the given provider
+        /// and provide a static image fallback (which is never cached).
+        /// </summary>
+        private static ICachedAvatarProvider WrapInCacheAndStaticFallback(IAvatarProvider rootProvider)
+        {
+            var persistentCacheProvider = new AvatarPersistentCache(rootProvider);
+            var memoryCachedProvider = new AvatarMemoryCache(persistentCacheProvider);
+
+            var providersWithCache = new ICachedAvatarProvider[]
+            {
+                persistentCacheProvider,
+                memoryCachedProvider,
+            };
+
+            var providerWithStaticFallback = new ChainedAvatarProvider(
+                memoryCachedProvider,
+                UserImageAvatarProvider);
+
+            return new CacheControlAvatarProvider(providerWithStaticFallback, providersWithCache);
```

Comment:
ChainedAvatarProvider combines providers.
AvatarPersistentCache combines providers.
AvatarMemoryCache combines providers.
CacheControlAvatarProvider combines providers.

I think we can make this code clearer.

Reply:
I'm not really seeing how `AvatarPersistentCache` and `AvatarMemoryCache` combine providers.
ChainedAvatarProvider - combines providers.
AvatarPersistentCache - a [decorator](https://en.wikipedia.org/wiki/Decorator_pattern) used for long term caching (file system), only takes a single provider
AvatarMemoryCache - a decorator with an in-memory cache (to reduce file system io overhead), only takes a single provider
CacheControlAvatarProvider - a kind of helper to orchestrate cache clearing.

The last one (CacheControlAvatarProvider) is probably the only one I could think of a few things that would make it cleaner, but I've no idea what changes (that wouldn't mix responsibilities) could make `ChainedAvatarProvider`, `AvatarPersistentCache` and `AvatarMemoryCache` clearer.

Reply:
rootProvider -> persistentCacheProvider -> memoryCachedProvider => providerWithStaticFallback ≡> CacheControlAvatarProvider

And caching management is woven into it.

> AvatarPersistentCache - a decorator

I agree, but managing caching is not relevant to getting an avatar.
This is related to the previous comment.

Could we split this into two **stateless** services?

Reply:
IMO caching should be something that is, in the best case, completely transparent to the consumer that is calling the interface. Currently the responsibilities are a bit mixed with `AvatarService.Default` - Is that the point you meant? Because if so, I can certainly fix that.

We could split the responsibilities into `AvatarService.DefaultProvider` (IAvatarProvider) and `AvatarService.CacheControl` (a new interface that would allow to delete the cache, which only the settings dialog would use for the "clear cache" button).

About the classes themselves, I'm sorry but I don't know what a **stateless** cache would even look like. I think to cache something it would always require some form of state unless I'm missing something.

Reply:
> completely transparent to the consumer that is calling the interface

Why then are we using the ICachedAvatarProvider interface?
Could we use IAvatarProvider for consumer?

> About the classes themselves, I'm sorry but I don't know what a stateless cache would even look like.

We add it to the IAvatarProvider, thereby breaking the original behavior.

`private static ICachedAvatarProvider WrapInCacheAndStaticFallback(IAvatarProvider rootProvider)`

to

`private static IAvatarProvider WrapInCacheAndStaticFallback(IAvatarProvider rootProvider)`

If we can do this, I think it will be easier to understand the idea further.

Reply:
@ivangrek can this be done in a follow up PR?

Reply:
I'd like to fix it with this PR, splitting the interfaces shouldn't be that hard, just didn't get what ivangrek meant at first.

Reply:
I made a few changes and split the interface into its responsibilities. Please let me know if the implementation has come closer to what you had in mind. I'm currently not so happy with the name `ICacheClearer` but can't think of a better one.

Reply:
Thanks. It looks better.

```
private static (IAvatarProvider provider, ICacheClearer cacheControl) SetupCachingAndFallback()
{
    var mruCache = new MruCache();
    var memoryAvatarProvider = new MemoryAvatarProvider(mruCache); // IAvatarProvider from memory
    
    var fileSystem = new FileSystem();
    var fileSystemAvatarProvider = new FileSystemAvatarProvider(fileSystem); // IAvatarProvider from file system
    
    var mainProvider = new ChainedAvatarProvider(
        memoryAvatarProvider,
        fileSystemAvatarProvider,
        HotSwapProvider,
        UserImageAvatarProvider);
        
    var avatarCache = new AvatarCache(mruCache, fileSystem);
    
    return (mainProvider, avatarCache);
}
```

ICacheClearer we could replace with the method in AvatarService.
ChainedAvatarProvider we could replace with the method in AvatarService.

```
var image = memoryAvatarProvider.GetAvatar(key)
    ?? fileSystemAvatarProvider.GetAvatar(key)
    ?? HotSwapProvider.GetAvatar(key)
    ?? UserImageAvatarProvider.GetAvatar(key);
    
return image;
```

If you think this is unnecessary, let me know, I will close the review.